SR-5739 Add refactoring action to collapse nested if

willtellis opened the pull request:
This implements a refactoring action in the IDE to collapse nested `if` statements. It handles collapsing only a single outer `if` statement with single inner `if` statement. Each `if` statement may contain one or more conditionals.

For example
```
if a > 2 {
  if a < 10 {}
}
```
would be refactored to
```
if a > 2, a < 10 {}
```

To do:
- [x] Filter out statements where the outer `if` statement contains anything besides the inner `if`
- [x] Handle collapsing `if let` statements
- [x] Add tests

Addresses [SR-5739](https://bugs.swift.org/browse/SR-5739).

willtellis committed:
Add test file for nested if refactoring action

willtellis committed:
[IDE] Add refactoring action to collapse nested if

This is a work-in-progress that adds a refactoring action to
collapse nested if statements. See SR-5739.

willtellis commented:
@harlanhaskins Thanks for offering to help me with this.

harlanhaskins commented on lib/IDE/Refactoring.cpp:
```diff
@@ -1463,6 +1463,64 @@ bool RefactoringActionExtractRepeatedExpr::performChange() {
+        SourceLoc StartLoc;
+        CollapsibleNestedIfInfo IfInfo;
+        IfStmtFinder(SourceLoc StartLoc): StartLoc(StartLoc), IfInfo() {}
+        bool walkToStmtPre(Stmt *S) {
```
Nitpick: LLVM style prefers early returns, rather than nesting the body of the function. This would be better expressed by:

```
if (S->getKind() != StmtKind::If) {
  return false;
}
```

harlanhaskins replied:
Nitpick: LLVM style uses 2 spaces for indentation.

CodaFi replied:
Could also dyn_cast first

```c++
auto *IFS = dyn_cast<IfStmt>(S);
if (!IFS) return false;
/**/ 
```

harlanhaskins commented on lib/IDE/Refactoring.cpp:
```diff
@@ -1463,6 +1463,64 @@ bool RefactoringActionExtractRepeatedExpr::performChange() {
+                    IfInfo.OuterIf = dyn_cast<IfStmt>(S);
+                    return true;
+                }
+                if (S->getKind() == StmtKind::If) {
```
This condition is the same as the outer condition, you can remove this if check

harlanhaskins commented:
Wrote a couple of nitpicks, but I think this is definitely in the right track! Thanks for tackling this and thanks for contributing! 🎉

willtellis committed:
Add unit test for refactoring action availability

nkcsgexi commented:
@willtellis Awesome work so far! Thank you for contributing to Swift refactoring.

willtellis committed:
Add refactoring test, move refactorability test

willtellis committed:
Remove unnecessary if statements, improve styling

harlanhaskins commented on lib/IDE/Refactoring.cpp:
```diff
@@ -1463,6 +1463,64 @@ bool RefactoringActionExtractRepeatedExpr::performChange() {
+    auto OuterIfConditionals = Target.OuterIf->getCond().vec();
+    auto InnerIfConditionals = Target.InnerIf->getCond().vec();
+
+    auto OuterIfConditionalText = Lexer::getCharSourceRangeFromSourceRange(SM, OuterIfConditionals[0].getSourceRange()).str();
```
Nitpick: 80 characters per line, please

nkcsgexi commented on test/refactoring/RefactoringKind/basic.swift:
```diff
@@ -174,6 +174,13 @@ func testStringLiteral() -> String {
+    if a < 10 {}
+  }
+}
+
```
can we add a test case to ensure such refactoring kind doesn't show up in a multi-statement brace statement? 
e.g 
```
if a > 2 { // not here.
  if a < 5 {
  }
  if a < 10 {
  }
}
```

nkcsgexi replied:
oh, I saw your listed to do above. Please ignore this.

willtellis committed:
Filter if statements with multiple conditionals

willtellis committed:
Add outer if filtering

Filter out statements where the outer if statement contains anything besides the inner if

mgrebenets commented on lib/IDE/Refactoring.cpp:
```diff
@@ -1463,6 +1463,112 @@ bool RefactoringActionExtractRepeatedExpr::performChange() {
 
+struct CollapsibleNestedIfInfo {
+  IfStmt *OuterIf;
+  IfStmt *InnerIf;
```
Does this mean that refactoring will support only 2 if statements initially?

nkcsgexi replied:
I'm personally ok with collapsing only 2 if statements at one refactoring operation, since the users can always do more by invoking the tool multiple times.

mgrebenets commented on lib/IDE/Refactoring.cpp:
```diff
@@ -1463,6 +1463,112 @@ bool RefactoringActionExtractRepeatedExpr::performChange() {
+
+    llvm::SmallString<64> DeclBuffer;
+    llvm::raw_svector_ostream OS(DeclBuffer);
+    OS << tok::kw_if << " (" << OuterIfConditionalText << ") && (";
```
While thinking on how multiple if-statement conditions could be combined together, I have re-discovered once again that in C-based languages comma is an operator and it's computed left-to-right, which means for boolean conditions it's same as `&&`.

So if we have

```swift
if a > 1 {
  if a < 1 || a > 3 {
  }
}
```

instead of `if (a > 1) && (a < 1 || a > 3)` we can just do `if a > 1, a < 1 || a > 3`.
It may look a bit off, but when we have optional binding of pattern matching inside if statements, then `,` works for all cases:

```
if let a = a, a > 1 {
  if let b = b, b < 3 || b > 5 {
    if d == false {
    }
  }
}
```

with `,` combining conditions is very easy:

```swift
if let a = a, a > 1, let b = b, b < 3 || b > 5, d == false {
}
```

willtellis replied:
I was planning on implementing combining of conditionals for `if let` statements using this approach next. If we prefer combining conditionals using commas for all cases, it certainly simplifies the logic. I believe it also makes it so we can easily combine `if` statements with more than one conditional. Thoughts?

willtellis replied:
Actually I just went ahead and did it. Let me know what you think!

willtellis committed:
Update unit tests for comma-separated conditionals

willtellis committed:
Collapse comma-separated conditionals

Make collapse nested if refactoring action separate conditionals with commas.
Update refactoring to handle collapsing if statements with multiple conditionals.
This has the consequence of automatically handling collapsing "let" and "case" conditionals.

willtellis commented:
I'm still dealing with one small issue. How can I apply proper indentation to my collapsed `if` statement? Currently,
```
if a > 2 {
  if a < 10 {
  }
}
```
becomes
```
if a > 2, a < 10 {
  }
```

harlanhaskins commented:
@nkcsgexi might have a better answer, but what if you used the column of the inner `if`?

This will break if, say, the code is

```
if x { if y { } }
```

But that might be a case we don’t care too much about...

nkcsgexi commented:
@willtellis Source editor will take care of indentation after applying a refactoring, so here we need to only worry about generating the correct refactoring content.

willtellis commented:
@nkcsgexi That's good to know, thanks! In that case, I think this PR is ready for approval. Do we need to run these smoke tests next?

harlanhaskins commented:
@swift-ci please smoke test

nkcsgexi commented:
@swift-ci Please test Linux platform

nkcsgexi commented:
@willtellis Merged!

harlanhaskins commented:
Thanks for the great contribution, @willtellis!

willtellis commented:
Thanks for all the help!
